File: Dot examples/ASP Book Examples/Book Examples/Chapter30/CS/App_Code/Product.cs
using System;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace myControls
{
    public class Product : CompositeControl
    {
        private ITemplate _itemTemplate;
        private ProductItem _item;

        public string Name
        {
            get
            {
                EnsureChildControls();
                return _item.Name;
            }
            set
            {
                EnsureChildControls();
                _item.Name = value;
            }
        }

        public Decimal Price
        {
            get
            {
                EnsureChildControls();
                return _item.Price;
            }
            set
            {
                EnsureChildControls();
                _item.Price = value;
            }
        }

        [TemplateContainer(typeof(ProductItem))]
        [PersistenceMode(PersistenceMode.InnerProperty)]
        public ITemplate ItemTemplate
        {
            get { return _itemTemplate; }
            set { _itemTemplate = value; }
        }

        protected override void CreateChildControls()
        {
            _item = new ProductItem();
            _itemTemplate.InstantiateIn(_item);
            Controls.Add(_item);
        }
    }

    public class ProductItem : WebControl, IDataItemContainer
    {
        private string _name;
        private decimal _price;

        public string Name
        {
            get { return _name; }
            set { _name = value; }
        }

        public decimal Price
        {
            get { return _price; }
            set { _price = value; }
        }

        public object DataItem
        {
            get
            {
                return this;
            }
        }

        public int DataItemIndex
        {
            get { return 0; }
        }

        public int DisplayIndex
        {
            get { return 0; }
        }
    }


}

File: Dot examples/ASP Book Examples/Book Examples/Chapter30/VB/App_Code/Product.vb
Imports System
Imports System.Web
Imports System.Web.UI
Imports System.Web.UI.WebControls

Namespace myControls

    Public Class Product
        Inherits CompositeControl

        Private _itemTemplate As ITemplate
        Private _item As ProductItem

        Public Property Name() As String
            Get
                EnsureChildControls()
                Return _item.Name
            End Get
            Set(ByVal Value As String)
                EnsureChildControls()
                _item.Name = Value
            End Set
        End Property

        Public Property Price() As Decimal
            Get
                EnsureChildControls()
                Return _item.Price
            End Get
            Set(ByVal Value As Decimal)
                EnsureChildControls()
                _item.Price = Value
            End Set
        End Property

        <TemplateContainer(GetType(ProductItem))> _
        <PersistenceMode(PersistenceMode.InnerProperty)> _
        Public Property ItemTemplate() As ITemplate
            Get
                Return _itemTemplate
            End Get
            Set(ByVal Value As ITemplate)
                _itemTemplate = Value
            End Set
        End Property

        Protected Overrides Sub CreateChildControls()
            _item = New ProductItem()
            _itemTemplate.InstantiateIn(_item)
            Controls.Add(_item)
        End Sub
    End Class

    Public Class ProductItem
        Inherits WebControl
        Implements IDataItemContainer

        Private _name As String
        Private _price As Decimal

        Public Property Name() As String
            Get
                Return _name
            End Get
            Set(ByVal Value As String)
                _name = Value
            End Set
        End Property

        Public Property Price() As Decimal
            Get
                Return _price
            End Get
            Set(ByVal Value As Decimal)
                _price = Value
            End Set
        End Property

        Public ReadOnly Property DataItem() As Object Implements IDataItemContainer.DataItem
            Get
                Return Me
            End Get
        End Property

        Public ReadOnly Property DataItemIndex() As Integer Implements IDataItemContainer.DataItemIndex
            Get
                Return 0
            End Get
        End Property

        Public ReadOnly Property DisplayIndex() As Integer Implements IDataItemContainer.DisplayIndex
            Get
                Return 0
            End Get
        End Property
    End Class


End Namespace
